### PR TITLE
add method

### PR DESCRIPTION
**Pull request by sanjupriya-manikyala**

### CODE REVIEW

File: Controllers/ProductController.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        public async Task<IActionResult> AddAsync([FromBody] ProductDTO product)
+        {
+              try
+                {
+                    var result = await _productService.AddAsync(product);
+                    if (result != null)
+                    {
+                        return CreatedAtAction("Get", new { id = result.Id }, result);
+                    }
+                        return NoContent();
```
Comment (rohitvipin) [inline Controllers/ProductController.cs]: dont do NoContent when it failed to insert. 
this is an error case and the end user need to know.
return 422 and let the user know it failed.

File: Controllers/ProductController.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+                }
+                catch (Exception ex)
+                {
+                    //_logger.LogInformation(StatusCodes.Status422UnprocessableEntity, ex.Message);
+                    return StatusCode(StatusCodes.Status422UnprocessableEntity, ex.Message);
+                }
```
Comment (rohitvipin) [inline Controllers/ProductController.cs:42]: generic exception when caught should return 5xx errors. like 500 or 503

File: CoreServices.csproj [hunk 3]
```diff
@@ -4,5 +4,22 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (rohitvipin) [inline CoreServices.csproj]: this item group is not required.

File: Models/ProductDBContext.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.HasAnnotation("Relational:Collation", "SQL_Latin1_General_CP1_CI_AS");
+
+            modelBuilder.Entity<Product>(entity =>
+            {
+                entity.ToTable("product");
+
+                entity.Property(e => e.Id).HasColumnName("ID");
```
Comment (rohitvipin) [inline Models/ProductDBContext.cs]: need to mention it is identity and auto generated

File: Models/ProductDBContext.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+                entity.Property(e => e.Name)
```
Comment (rohitvipin) [inline Models/ProductDBContext.cs:40]: should be marked required

File: Services/ProductService.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        private readonly IRepository _repository;
+
+        public ProductService(IRepository repository)
+        {
+            _repository = repository;
+        }
+
+        public async Task<ProductDTO> AddAsync(ProductDTO product)
+        {
+            var model = new Product()
```
Comment (rohitvipin) [inline Services/ProductService.cs]: () not required

File: Services/ProductService.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            {
+                Name = product.Name,
+                Price = product.Price
+            };
+            var result = await _repository.AddAsync(model);
+            {
```
Comment (rohitvipin) [inline Services/ProductService.cs:25]: what is this parenthesis for?

File: Services/ProductService.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                product.Id = result.Id;
+                product.Name = result.Name;
```
Comment (rohitvipin) [inline Services/ProductService.cs:27]: i would prefer a mapper or helper but i think its ok

File: Startup.cs [hunk 9]
```diff
@@ -26,6 +31,14 @@ public Startup(IConfiguration configuration)
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddControllers();
+            services.AddDbContext<ProductDBContext>(item => item.UseSqlServer(Configuration.GetConnectionString("ProductDBConnection")));
+            services.AddTransient<IRepository, ProductRepository>();
```
Comment (rohitvipin) [inline Startup.cs]: AddTransient -> AddScoped

File: Startup.cs [hunk 10]
```diff
@@ -44,7 +57,17 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
             app.UseEndpoints(endpoints =>
             {
-                endpoints.MapControllers();
+                endpoints.MapControllerRoute(
+                    name: "default",
+                    pattern: "{controller=Home}/{action=Index}/{id?}");
```
Comment (rohitvipin) [inline Startup.cs]: i think this is mvc and not web API 
use map controllers 

 app.UseEndpoints(endpoints => endpoints.MapControllers());

**Review by rohitvipin — CHANGES_REQUESTED** [review]

File: Controllers/ProductController.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+using System.Threading.Tasks;
+
+namespace CoreServices.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class ProductController : ControllerBase
+    {
+        private readonly ProductService _productService;
+        //private readonly ILogger _logger;
```
Comment (rohitvipin) [inline Controllers/ProductController.cs]: why commented?

File: Models/Product.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+﻿
+#nullable disable
+
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace CoreServices.Models
+{
+    public partial class Product
+    {
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
```
Comment (rohitvipin) [inline Models/Product.cs]: stick to one, either attributes or fluent way of defining database constraints

File: Models/ProductDBContext.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.HasAnnotation("Relational:Collation", "SQL_Latin1_General_CP1_CI_AS");
+
+            modelBuilder.Entity<Product>(entity =>
+            {
+                entity.ToTable("product");
+
+                entity.Property(e => e.Id).HasColumnName("ID").ValueGeneratedOnAdd();
```
Comment (rohitvipin) [inline Models/ProductDBContext.cs:38]: you can use HasKey

File: ProductControllerTests.cs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        public async void Task_AddAsync_ValidData_Return_OkResult()
+        {
+            //Arrange
+            var product = new ProductDTO() { Id = 5, Name = "Test Name5", Price = 8120 };
+
+            //Act  
+            var data = await _productController.AddAsync(product);
+
+            //Assert  
+            Assert.IsType<OkObjectResult>(data);
```
Comment (rohitvipin) [inline ProductControllerTests.cs]: just asserting on type in not good enough for a test

File: ProductControllerTests.cs [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            //Arrange
+            var product = new ProductDTO() { Id = 5, Name = "Test Name5", Price = 8120 };
+            _productController.StatusCode(StatusCodes.Status500InternalServerError);
+
+            //Act  
+            var data = await _productController.AddAsync(product) as CreatedAtActionResult;
+
+            //Assert  
+            Assert.Equal(StatusCodes.Status500InternalServerError, data.StatusCode);
```
Comment (rohitvipin) [inline ProductControllerTests.cs]: why would a created at action return error?

File: ProductControllerTests.cs [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+            //Assert  
+            Assert.Equal(product, value);
+        }
+
+        [Fact]
+        public void Add_ValidDataPassed_Return_MatchResult()
+        {
+            //Arrange  
+            var factory = new ConnectionFactory();
+            var context = factory.CreateContextForInMemory();
```
Comment (rohitvipin) [inline ProductControllerTests.cs]: you are mocking the service, why would you need context and in memory database?

File: ProductControllerTests.cs [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Moq;
+using Xunit;
+
+namespace CoreServices.Tests
+{
+    public class ProductControllerTests
+    {
+        private readonly ProductController _productController;
+        private readonly Mock<ProductService> _mockProductService;
```
Comment (rohitvipin) [inline ProductControllerTests.cs:20]: where are you using the mock?

File: Services/ProductService.cs [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+using CoreServices.Repository;
+using System.Threading.Tasks;
+
+namespace CoreServices.Services
+{
+    public class ProductService
+    {
+        private readonly IRepository _repository;
+
+        public ProductService(IRepository repository)
```
Comment (rohitvipin) [inline Services/ProductService.cs:12]: Where are tests for this class?

**Review by rohitvipin — CHANGES_REQUESTED** [review]

File: Controllers/ProductController.cs [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+                    if (result != null)
+                    {
+                        return CreatedAtAction("Get", new { id = result.Id }, result);
+                    }
+                        return StatusCode(StatusCodes.Status422UnprocessableEntity);
+
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogInformation(StatusCodes.Status422UnprocessableEntity, ex.Message);
```
Comment (rohitvipin) [inline Controllers/ProductController.cs]: _logger.LogInformation(StatusCodes.Status422UnprocessableEntity, ex.Message);???
It should be something like _logger.LogError(ex);

File: ProductControllerTests.cs [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+            //Arrange
+            var product = new ProductDTO() { Id = 5, Name = "Test Name5", Price = 8120 };
+            _mockProductService.Setup(p => p.AddAsync(product)).ReturnsAsync(product);
+
+            //Act  
+            var data = await _productController.AddAsync(product);
+
+            //Assert
+            Assert.IsType<CreatedAtActionResult>(data);
+            Assert.IsAssignableFrom<CreatedAtActionResult>(data);
```
Comment (rohitvipin) [inline ProductControllerTests.cs]: needs something that would check Assert.Equal(expected, actual) or using FluentAssertions

File: ProductControllerTests.cs [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+            var data = await _productController.AddAsync(product) as CreatedAtActionResult;
+            var value = data.Value as ProductDTO;
+
+            //Assert
+            Assert.NotNull(value);
+            Assert.Equal(product, value);
+        }
+
+        [Fact]
+        public async void Task_AddAsync_PassedInValidData_Return_BadRequest()
```
Comment (rohitvipin) [inline ProductControllerTests.cs:72]: test for StatusCodes.Status500InternalServerError is missing

**Review by rohitvipin — CHANGES_REQUESTED** [review]

File: ProductControllerTests.cs [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+            //Arrange
+            var product = new ProductDTO() { Id = 5, Name = "Test Name5", Price = 8120 };
+            _mockProductService.Setup(p => p.AddAsync(product)).ReturnsAsync(product);
+
+            //Act  
+            var data = await _productController.AddAsync(product);
+
+            //Assert
+            var resulttype = Assert.IsType<CreatedAtActionResult>(data);
+            Assert.Equal(typeof(CreatedAtActionResult), resulttype.GetType());
```
Comment (rohitvipin) [inline ProductControllerTests.cs]: you are only checking type. 
to check type you can use is type or is assignable from. 
you need to assert on actual data

File: ProductControllerTests.cs [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+            _mockProductService = new Mock<ProductService>(_repository.Object);
+            _productController = new ProductController(_mockProductService.Object, _logger);
+        }
+
+
+        [Fact]
+        public async void Task_AddAsync_PassedValidData_Return_OkResult()
+        {
+            //Arrange
+            var product = new ProductDTO() { Id = 5, Name = "Test Name5", Price = 8120 };
```
Comment (janani-ravikumar) [inline ProductControllerTests.cs]: Can you use fixture to create test data?

File: ProductControllerTests.cs (hunk 22 above)
Comment (janani-ravikumar) [inline ProductControllerTests.cs]: To validate result type, the below one is fine
Assert.IsType<CreatedAtActionResult<CreatedAtActionResult>>(data);
Please remove
 Assert.Equal(typeof(CreatedAtActionResult), resulttype.GetType());

Comment (janani-ravikumar) [inline ProductControllerTests.cs]: Please validate the returned content

File: ProductControllerTests.cs [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+            //Assert
+            var resulttype = Assert.IsType<CreatedAtActionResult>(data);
+            Assert.Equal(typeof(CreatedAtActionResult), resulttype.GetType());
+        }
+
+        [Fact]
+        public async void Task_AddAsync_PassedValidData_Return_MatchResult()
+        {
+            //Arrange
+            var product = new ProductDTO() { Id = 5, Name = "Test Name5", Price = 8120 };
```
Comment (janani-ravikumar) [inline ProductControllerTests.cs]: Can you use fixture to create test data?

File: ProductControllerTests.cs [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+            //Arrange
+            var product = new ProductDTO { Id = 5, Name = "Test Name5", Price = 8120 };
+            _mockProductService.Setup(p => p.AddAsync(product)).ReturnsAsync(product);
+            _productController.StatusCode(StatusCodes.Status500InternalServerError);
+
+            //Act  
+            var data = await _productController.AddAsync(product: null);
+
+            //Assert
+            var resulttype = Assert.IsType<StatusCodeResult>(data);
```
Comment (janani-ravikumar) [inline ProductControllerTests.cs]: Please use camelCase for variables

File: ProductRepositoryTests.cs [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            //Arrange
+            var product = new Product { Id = 6, Name = "Test Name6", Price = 670 };
+            _dbContext.Setup(p => p.Products.Add(product));
+
+            //Act
+            var data = await _productRepository.AddAsync(product);
+
+            //Assert
+            Assert.Equal(product, data);
```
Comment (janani-ravikumar) [inline ProductRepositoryTests.cs]: Please validate the return type

File: ProductServiceTests.cs [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+            //Arrange
+            var dto = new ProductDTO() { Id = 6, Name = "Test Name6", Price = 670 };
+            var product = new Product { Id = dto.Id, Name = dto.Name, Price = dto.Price };
+            _repository.Setup(p => p.AddAsync(product)).ReturnsAsync(product);
+
+            //Act
+            var data = await _productService.AddAsync(dto);
+
+            //Assert
+            Assert.Equal(dto, data);
```
Comment (janani-ravikumar) [inline ProductServiceTests.cs:66]: Please validate the return type

File: ProductServiceTests.cs [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+            var fixture = new Fixture();
+            var product = fixture.Create<Product>();
+            var dto = fixture.Create<ProductDTO>();
+            _repository.Setup(p => p.AddAsync(product)).ReturnsAsync(product);
+
+            //Act
+            var data = await _productService.AddAsync(dto);
+
+            //Assert
+            Assert.IsType<CreatedAtActionResult>(data);
```
Comment (janani-ravikumar) [inline ProductServiceTests.cs]: Service method is not returning CreatedAtActionResult, please check

File: ProductServiceTests.cs [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+            //Arrange
+            var fixture = new Fixture();
+            var product = fixture.Create<Product>();
+            var dto = fixture.Create<ProductDTO>();
+            _repository.Setup(p => p.AddAsync(product)).ReturnsAsync(product);
+
+            //Act
+            var data = await _productService.AddAsync(dto);
+            data = null;
+
```
Comment (janani-ravikumar) [inline ProductServiceTests.cs:85]: Why you are assigning null here?

File: ProductControllerTests.cs [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+        private readonly ProductController _productController;
+        private readonly ILogger _logger;
+        private readonly Mock<IRepository> _repository;
+        private readonly Mock<ProductService> _mockProductService;
+
+        public ProductControllerTests()
+        {
+            _logger = new Mock<ILogger>().Object;
+            _repository = new Mock<IRepository>();
+            _mockProductService = new Mock<ProductService>(_repository.Object);
```
Comment (janani-ravikumar) [inline ProductControllerTests.cs:26]: _repository = new Mock<IRepository>();
_repository is not needed.
 _mockProductService = new Mock<ProductService<ProductService>>(); --> This should be fine

File: Controllers/ProductController.cs [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+        {
+              try
+                {
+                    var result = await _productService.AddAsync(product);
+                    if (result != null)
+                    {
+                        return CreatedAtAction("Get", new { id = result.Id }, result);
+                    }
+                        return StatusCode(StatusCodes.Status422UnprocessableEntity);
+                }
```
Comment (janani-ravikumar) [inline Controllers/ProductController.cs:33]: Please format your code

File: Controllers/ProductController.cs [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+                var result = await _productService.AddAsync(product);
+                if (result != null)
+                {
+                    return CreatedAtAction("Get", new { id = result.Id }, result);
+                }
+                return StatusCode(StatusCodes.Status422UnprocessableEntity);
+              }
+              catch (Exception ex)
+              {
+                    _logger.LogError(StatusCodes.Status422UnprocessableEntity, ex.Message);
```
Comment (rohitvipin) [inline Controllers/ProductController.cs]: _logger.LogError(StatusCodes.Status422UnprocessableEntity, ex.Message);
is wrong

File: Models/ProductDBContext.cs (hunk 13 above)
Comment (rohitvipin) [reply Models/ProductDBContext.cs:38]: Why is the column name differnt?

**Review by rohitvipin — CHANGES_REQUESTED** [review]

File: ProductControllerTests.cs (hunk 22 above)
Comment (sanjupriya-manikyala) [reply ProductControllerTests.cs]: yes, this is fixed in the later versions

File: Controllers/ProductController.cs (hunk 32 above)
Comment (sanjupriya-manikyala) [reply Controllers/ProductController.cs]: will change

**Comment by sanjupriya-manikyala** [comment]
will correct it

**Pull request merged**